Updates to build on OS X

boneill42 opened the pull request:
I need some help here.. it's been about 15 years since I did C/POSIX.

I don't know the implications of NOT setting the process scheduler, but it appears as those OS X does not have the function:
https://www.gnu.org/software/gnulib/manual/html_node/sched_005fsetscheduler.html

For now, I used preprocessing to exclude this from the APPLE build.

boneill42 committed:
Updates to preprocessiong and make files to build start-stop-daemon on OS X

richardkchapman commented on tools/start-stop-daemon/start-stop-daemon.c:
```diff
@@ -39,6 +39,8 @@
 #elif defined(__NetBSD__)
 #  define OSNetBSD
+#elif defined(__APPLE__)
+#  define OSLinux
```
I would have thought apple was closer to OSFreeBSD than OSLinux

boneill42 replied:
I thought the same thing.  When using OSFreeBSD, I got the following incompatibilities:

In file included from /Users/bone/git/boneill42/hpcc/tools/start-stop-daemon/start-stop-daemon.c:65:
/usr/include/sys/proc.h:118:19: error: field has incomplete type 'struct itimerval'
        struct  itimerval p_realtimer;  /* Alarm timer. */
                          ^
/usr/include/sys/proc.h:118:9: note: forward declaration of 'struct itimerval'
        struct  itimerval p_realtimer;  /* Alarm timer. */
                ^
/Users/bone/git/boneill42/hpcc/tools/start-stop-daemon/start-stop-daemon.c:1263:9: warning: implicit declaration of function 'do_procinit' is invalid in C99 [-Wimplicit-function-declaration]
        do_procinit();

boneill42 replied:
(which is why I punted and went with OSLinux)

richardkchapman commented:
The effect will be that the -P option will not work:

       -P, --procsched policy:priority
              This alters the process scheduler policy and priority of the process before starting it. The priority can be optionally specified by appending a : followed by the value.
              The default priority is 0. The currently supported policy values are other, fifo and rr.

AFAIK we don't use that option in any of our usage of this program (though I'm not really very familiar with how we use it - @Michael-Gardner or @xwang2713 may know more).

Michael-Gardner commented:
We don't use the -P option at all for start-stop-daemon.  It shouldn't be an issue if OS X doesn't support it.

boneill42 committed:
Updates to detail dependencies required for Linux build.
(used base Ubuntu 14 to test)

richardkchapman commented on tools/CMakeLists.txt:
```diff
@@ -39,3 +39,7 @@ if ( CMAKE_SYSTEM MATCHES Linux )
     message ("-- Building start-stop-daemon")
     HPCC_ADD_SUBDIRECTORY (start-stop-daemon)
 endif()
+
```
Would be preferable to say

```
  if ( CMAKE_SYSTEM MATCHES "Linux|APPLE" )
```
above (and at the same time update the comment that says start-stop-daemon is linux only)

boneill42 replied:
Agreed, will do.

boneill42 committed:
Merge branch 'master' of github.com:hpcc-systems/HPCC-Platform

boneill42 commented:
Just a heads up, I'm pushing another change to this commit.  jfile.cpp had an issue on OS X, which resulted in a failed compile:

W20150402-123033 failed
<Result>
 <Exception><Code>4294967295</Code><Source>Roxie</Source><Message>UNIMPLEMENTED feature at /Users/bone/git/boneill42/hpcc/system/jlib/jfile.cpp(474)</Message></Exception>
</Result>

boneill42 committed:
ECL compiles were failing on OS X, with:

<Result>
 <Exception><Code>4294967295</Code><Source>Roxie</Source><Message>UNIMPLEMENTED feature at /Users/bone/git/boneill42/hpcc/system/jlib/jfile.cpp(474)</Message></Exception>
</Result>

This commit fixes jfile, making the implementation the same as Linux.

Also -- addresses @richardkchapman's comments on the pull request.

boneill42 committed:
Merge branch 'master' of github.com:boneill42/HPCC-Platform

boneill42 commented on system/jlib/jfile.cpp:
```diff
@@ -470,8 +472,6 @@ bool CFile::setTime(const CDateTime * createTime, const CDateTime * modifiedTime
     if (modifiedTime)
         am.modtime  = timetFromIDateTime (modifiedTime);
     return (utime(filename, &am)==0);
-#else
```
NOTE: This is where the issue was -- was failing the ECL compile because this was UNIMPLEMENTED for APPLE

boneill42 committed:
Resolves an issue when compiling ECL on OS X.

@loader_path is not available in this context, which means the library cannot load.

This results in an error:

<Result>
 <Exception><Source>Roxie</Source><Message>Error loading /var/lib/HPCCSystems/queries/myroxie/libW20150402-131125.dylib: dlopen(/var/lib/HPCCSystems/queries/myroxie/libW20150402-131125.dylib, 2): Library not loaded: @loader_path/../lib/libeclrtl.dylib
  Referenced from: /var/lib/HPCCSystems/queries/myroxie/libW20150402-131125.dylib
  Reason: image not found</Message></Exception>
</Result>

This fixes that error and allows the ECL->C to link properly.

This final commit allows the helloworld.ecl to pass on OS X.
(running esp, dali, eclcc, and roxie)

boneill42 commented:
OK, with the final change to commonSetup.cmake, I am able to run esp, eclcc, dali and roxie on OS X.  I can issue the helloworld.ecl, and receive a valid result!

boneill42 commented:
Update: the plugins work as well, but I needed to copy the libraries into /opt/HPCCSystems/lib.

richardkchapman commented:
Worth opening a JIRA for the plugins issue. I can take a look at that (or you can if you get to it first).
